### DISCUSSION TITLE
Improve FAKE Support TODO List (discussion and feedback)

### DISCUSSION THREAD

**Discussion by matthid** [ionide/ionide-vscode-fsharp#1144] [Ideas]
Here is what I'd like to see in Ionide and potentially contribute myself the next couple of weeks. This is more a TODO list for myself but I'd like to see people step in and help with some of the features. Feedback suggestions or criticism (don't do X, do Y before X, do Y instead of X) is welcome. The current order is assumed to be:

- [ ] Fix ![image](https://user-images.githubusercontent.com/1866463/60402820-ef135f00-9b94-11e9-8e8a-f7156c0a0096.png)
- [x] Fix write-info results empty scripts (no file created)
![image](https://user-images.githubusercontent.com/1866463/60746497-48a5d000-9f7f-11e9-8ed7-7d757b2b06c5.png)
- [ ] Detect if dependencies have changed and popup the user to update its dependencies. Run `paket install` in case of external `paket.dependencies`, delete `<build>.fsx.lock` otherwise
- [x] Fix issue with Ionide changing `Paket.Restore.targets` -> https://github.com/fsharp/FsAutoComplete/pull/417
- [ ] Fix issue with Ionide 4.0 regressed on https://github.com/fsharp/FsAutoComplete/pull/273 (ie it shows a warning on `#r "paket:`)
- [ ] Replace the "Run F# Script" with a "Run FAKE Script" button when we detect that the current file is a FAKE script (and start FAKE instead of fsi). Same with "Debug".
- [ ] ~~Add proper VS-Code-commands to execute Fake~~ apparently [part of the next point](https://github.com/ionide/ionide-vscode-fsharp/issues/1144#issuecomment-505031295) 
- [ ] Add proper VS-Code integration of VS-Code-Tasks

  > The one thing we may want to add on Ionide side is implementing TaskProvider - https://code.visualstudio.com/api/extension-guides/task-provider

- [ ] Automatically refresh the FAKE targets outline on changes
- [ ] Add some content to http://ionide.io/docs/fake/
- [ ] Improve visibility when background jobs are running
- [ ] Remove the FAKE targets outline for regular F# scripts
- [ ] Target visualization (https://twitter.com/matthi__d/status/1138403342516989955)
      Current idea is to have some "preview" feature where you can select a target and then get a visualization of what FAKE would run and when.
- [ ] Quickfix for adding new packages (not 100% sure but I think I'll generate some "index" or dictionary from all code in the FAKE repository and then add quickfixes for the corresponding compiler errors if we know some symbol exists somewhere)
- [ ] Try to improve/check performance of the features.
- [ ] Add option to disable automatic FAKE script execution (only when user clicks the refresh button)
- [ ] Check why "Just My Code" setting is required to be off (https://github.com/ionide/ionide-vscode-fsharp/pull/1137#discussion_r294721820)

[a comment by an automated account is not shown]

**Comment by Krzysztof-Cieslak** [comment]
> Add proper VS-Code-commands to execute Fake

> Add proper VS-Code integration of VS-Code-Tasks

I think only one of those needs to be done (latter) - integration with Task API will integrate us with built-in ways of running tasks, no need for custom commands

**Comment by drk-mtr** [comment]
Potential idea: check for Fake 5 in the opened folder, and disable the `FAKE Targets Outline` option if it's found. Could also add a note to documentation to advise it's unsupported.

This could help avoid confusion where people:
- see `unknown error code [object Object]`
- assume they're doing something wrong
- read the `README.md` for `ionide-vscode-fake` and try and fail to create a `.ionide` file because there's already a folder with this name
- decide to create an issue... which wouldn't help since you guys are fully aware already ;)

Thanks for the amazing software!

**Comment by matthid** [comment]
@drk-mtr so you are taking about:

> [ ] Remove the FAKE targets outline for regular F# scripts

Correct?

**Comment by drk-mtr** [comment]
I believe so - I mean the collapsible "Fake Targets Outline" menu item that appears as an option in the Explorer panel. I'm not too sure what you mean by "regular F# scripts" - I think I interpreted that as ".fsx files" when I scanned through your list earlier, whereas this is a full project.
